src/Workspaces/Core/Portable/Execution/ChecksumTreeBuilder.cs

```diff
@@ -133,58 +134,96 @@ private Task<ChecksumCollection> BuildAsync(IReadOnlyList<AnalyzerReference> key
             return CreateChecksumCollectionsAsync(documentStates, kind, snapshotBuilder.BuildAsync, cancellationToken);
         }
 
-        private Task<ChecksumCollection> CreateChecksumCollectionsAsync(
+        private async Task<ChecksumCollection> CreateChecksumCollectionsAsync<TValue, TChecksumObject>(
+            IEnumerable<TValue> items, string kind, Func<TValue, CancellationToken, Task<TChecksumObject>> buildAsync, CancellationToken cancellationToken) where TChecksumObject : ChecksumObject
+        {
+            using (var pooledObject = Creator.CreateList<Task<TChecksumObject>>())
+            {
+                // create asyn checksums concurrently
+                var tasks = pooledObject.Object;
+                foreach (var item in items)
+                {
+                    tasks.Add(buildAsync(item, cancellationToken));
+                }
+
+                await Task.WhenAll(tasks).ConfigureAwait(false);
+
+                var checksums = new Checksum[tasks.Count];
+
+                for (var i = 0; i < tasks.Count; i++)
+                {
+                    var task = tasks[i];
+
+                    // task is already completed. so wait will return right away as long as there is no exception.
+                    // if there is exception, then it will throw it. if it is cancellation exception, it will make sure
+                    // it throws cancellation exception rather than aggregated exception.
+                    task.Wait(cancellationToken);
```

Comment:
@CyrusNajmabadi  @heejaechang  I happy with this compromise.
```c#
                for (var i = 0; i < tasks.Count; i++)
                {
                    var task = tasks[i];

                    // task is already completed. so wait will return right away as long as there is no exception.
                    // if there is exception, then it will throw it. if it is cancellation exception, it will make sure
                    // it throws cancellation exception rather than aggregated exception.
                    await task.ConfigureAwait( false );

                    // the task is not cancelled, get result. if we do .Result without Wait, this will throw
                    // aggregated exception for cancellation.
                    checksums[i] = task.Result.Checksum;
                }
```